File: lib/reference-policy.ts
import type { MessageRecord, ReferenceRecord } from "./category-repository";

export const REFERENCE_CURATION_STATUSES = ["pending", "accepted", "rejected", "archived"] as const;
export type ReferenceCurationStatus = typeof REFERENCE_CURATION_STATUSES[number];

export const REFERENCE_REJECTION_REASON_CODES = [
  "out_of_scope",
  "policy_exclusion",
  "duplicate",
  "low_quality",
  "unavailable",
  "provenance",
  "other",
] as const;
export type ReferenceRejectionReasonCode = typeof REFERENCE_REJECTION_REASON_CODES[number];

const REFERENCE_REJECTION_REASON_CODE_SET = new Set<string>(REFERENCE_REJECTION_REASON_CODES);
const SCOPE_TRAINING_NEGATIVE_REASON_CODE_SET = new Set<string>(["out_of_scope", "policy_exclusion"]);

export function normalizeReferenceCurationStatus(value: unknown, defaultStatus: ReferenceCurationStatus = "pending"): ReferenceCurationStatus {
  const normalized = normalizePolicyToken(value ?? defaultStatus);
  if (normalized === "accepted" || normalized === "accept" || normalized === "ready" || normalized === "trusted") return "accepted";
  if (normalized === "rejected" || normalized === "reject" || normalized === "discarded") return "rejected";
  if (normalized === "archived" || normalized === "archive") return "archived";
  return "pending";
}

export function referenceCurationStatusForAction(action: "accept" | "reject" | "reopen" | "archive"): ReferenceCurationStatus {
  if (action === "accept") return "accepted";
  if (action === "reject") return "rejected";
  if (action === "archive") return "archived";
  return "pending";
}

export function normalizeReferenceRejectionReasonCode(value: unknown): ReferenceRejectionReasonCode | null {
  const normalized = normalizePolicyToken(value);
  if (!normalized) return null;
  return REFERENCE_REJECTION_REASON_CODE_SET.has(normalized) ? normalized as ReferenceRejectionReasonCode : null;
}

export function isEvidenceEligibleReference(reference: Pick<ReferenceRecord, "versionState" | "curationStatus"> | null | undefined): boolean {
  return Boolean(reference)
    && reference?.versionState === "current"
    && normalizeReferenceCurationStatus(reference?.curationStatus) === "accepted";
}

export function referenceReasonCode(reference: Pick<ReferenceRecord, "metadata"> | null | undefined, messages: Array<Pick<MessageRecord, "metadata" | "createdAt">> = []): ReferenceRejectionReasonCode | null {
  const metadataReason = reasonCodeFromMetadata(reference?.metadata);
  if (metadataReason) return metadataReason;
  const sortedMessages = [...messages].sort((left, right) => String(right.createdAt ?? "").localeCompare(String(left.createdAt ?? "")));
  for (const message of sortedMessages) {
    const reason = reasonCodeFromMetadata(message.metadata);
    if (reason) return reason;
  }
  return null;
}

export function scopeTrainingLabelForReference(reference: Pick<ReferenceRecord, "versionState" | "curationStatus" | "metadata"> | null | undefined, messages: Array<Pick<MessageRecord, "metadata" | "createdAt">> = []): "positive" | "negative" | null {
  if (!reference || reference.versionState !== "current") return null;
  const status = normalizeReferenceCurationStatus(reference.curationStatus);
  if (status === "accepted") return "positive";
  if (status !== "rejected") return null;
  const reasonCode = referenceReasonCode(reference, messages);
  return reasonCode && SCOPE_TRAINING_NEGATIVE_REASON_CODE_SET.has(reasonCode) ? "negative" : null;
}

function reasonCodeFromMetadata(metadata: unknown): ReferenceRejectionReasonCode | null {
  const parsed = parseMetadataObject(metadata);
  return normalizeReferenceRejectionReasonCode(
    parsed.reasonCode
      ?? parsed.reason_code
      ?? parsed.curationReasonCode
      ?? parsed.curation_reason_code
      ?? parsed.rejectionReasonCode
      ?? parsed.rejection_reason_code,
  );
}

function parseMetadataObject(value: unknown): Record<string, unknown> {
  if (!value) return {};
  if (typeof value === "object" && !Array.isArray(value)) return value as Record<string, unknown>;
  if (typeof value !== "string") return {};
  try {
    const parsed = JSON.parse(value);
    return parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed as Record<string, unknown> : {};
  } catch {
    return {};
  }
}

function normalizePolicyToken(value: unknown): string {
  return String(value ?? "")
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_|_$/g, "");
}

File: src/papyrus_newsroom/reference_policy.py
from __future__ import annotations

import json
import re
from typing import Any


REFERENCE_CURATION_STATUSES = {"pending", "accepted", "rejected", "archived"}
REFERENCE_REJECTION_REASON_CODES = {
    "out_of_scope",
    "policy_exclusion",
    "duplicate",
    "low_quality",
    "unavailable",
    "provenance",
    "other",
}
SCOPE_TRAINING_NEGATIVE_REASON_CODES = {"out_of_scope", "policy_exclusion"}


def normalize_reference_curation_status(value: Any, default: str = "pending") -> str:
    normalized = _normalize_policy_token(value if value is not None else default)
    if normalized in {"accepted", "accept", "ready", "trusted"}:
        return "accepted"
    if normalized in {"rejected", "reject", "discarded"}:
        return "rejected"
    if normalized in {"archived", "archive"}:
        return "archived"
    return "pending"


def reference_curation_status_for_action(action: str) -> str:
    normalized = _normalize_policy_token(action)
    if normalized == "accept":
        return "accepted"
    if normalized == "reject":
        return "rejected"
    if normalized == "archive":
        return "archived"
    if normalized == "reopen":
        return "pending"
    raise ValueError(f"Unsupported reference curation action {action}.")


def normalize_reference_rejection_reason_code(value: Any) -> str | None:
    normalized = _normalize_policy_token(value)
    if not normalized:
        return None
    if normalized not in REFERENCE_REJECTION_REASON_CODES:
        raise ValueError(f"Unsupported reference rejection reason code {value}.")
    return normalized


def is_evidence_eligible_reference(reference: dict[str, Any] | None) -> bool:
    if not reference:
        return False
    return (
        reference.get("versionState") == "current"
        and normalize_reference_curation_status(reference.get("curationStatus")) == "accepted"
    )


def reference_reason_code(reference: dict[str, Any] | None, messages: list[dict[str, Any]] | None = None) -> str | None:
    metadata_reason = _reason_code_from_metadata((reference or {}).get("metadata"))
    if metadata_reason:
        return metadata_reason
    for message in sorted(messages or [], key=lambda entry: str(entry.get("createdAt") or ""), reverse=True):
        reason = _reason_code_from_metadata(message.get("metadata"))
        if reason:
            return reason
    return None


def scope_training_label_for_reference(reference: dict[str, Any] | None, messages: list[dict[str, Any]] | None = None) -> str | None:
    if not reference or reference.get("versionState") != "current":
        return None
    status = normalize_reference_curation_status(reference.get("curationStatus"))
    if status == "accepted":
        return "positive"
    if status != "rejected":
        return None
    reason_code = reference_reason_code(reference, messages)
    return "negative" if reason_code in SCOPE_TRAINING_NEGATIVE_REASON_CODES else None


def _reason_code_from_metadata(metadata: Any) -> str | None:
    parsed = _parse_metadata(metadata)
    return normalize_reference_rejection_reason_code(
        parsed.get("reasonCode")
        or parsed.get("reason_code")
        or parsed.get("curationReasonCode")
        or parsed.get("curation_reason_code")
        or parsed.get("rejectionReasonCode")
        or parsed.get("rejection_reason_code")
    )


def _parse_metadata(value: Any) -> dict[str, Any]:
    if not value:
        return {}
    if isinstance(value, dict):
        return value
    if not isinstance(value, str):
        return {}
    try:
        parsed = json.loads(value)
    except json.JSONDecodeError:
        return {}
    return parsed if isinstance(parsed, dict) else {}


def _normalize_policy_token(value: Any) -> str:
    return re.sub(r"(^_+|_+$)", "", re.sub(r"[^a-z0-9]+", "_", str(value or "").strip().lower()))
